### PR TITLE
Timeouts for `fetch install` and expose more metadata from downloaded packages

### PR DESCRIPTION
**Pull request by ditman**
* Expose popularity/maintenance/health scores for the downloaded packages.
  * Write the values to index.json so they can be used from other tools
* Add a `--timeout` option to 'fetch' command (disabled by default)
* Minor improvements to 'fetch':
  * Reduce output when not 'verbose'
  * Tweak argument parsing
    * Prevent -[no-]no-install flag
    * Other minor things

Use the new timeout like this:

```
dart pub_crawl.dart fetch --criteria flutter --max 5 --timeout 300
```

Now, if the pub get operation takes longer than 300 seconds, it'll be aborted and the script should continue downloading other packages.

### CODE REVIEW

**Comment by ditman** [comment]
Ooops! I'll fix the Travis breakages ASAP.

File: lib/src/cache.dart [hunk 1]
```diff
@@ -81,7 +82,21 @@ class Cache {
...
+    if (timeout <= 0) {
+      // Nothing to do here
+      return futureProcess;
+    }
+    // Wrap the futureProcess in timeout, and return it when it's ready (or throw if timeout exceeded!)
+    return futureProcess.then((Process p) =>
+        p.exitCode.timeout(Duration(seconds: timeout), onTimeout: () {
+          Process.killPid(p.pid);
+          throw Exception(
+              'Process with PID ${p.pid} took longer than ${timeout}s to complete. Killed.');
```
Comment (pq) [inline lib/src/cache.dart:95]: should this be 

    ${timeout} ms

?

File: lib/src/commands/fetch.dart [hunk 2]
```diff
@@ -121,7 +139,7 @@ class FetchCommand extends BaseCommand {
       print('(Max processed package count of $maxCount reached.)');
     }
 
-    print(count);
+    print('---');
```
Comment (pq) [inline lib/src/commands/fetch.dart:142]: IIRC printing the count was useful to get some sense for overall progress but maybe that's redundant now somehow?

File: lib/src/filters.dart [hunk 3]
```diff
@@ -2,5 +2,5 @@ import 'package:pub_semver/pub_semver.dart';
 
 final _dart2 = VersionConstraint.parse('>=2.0.0');
 
-bool isDart2(String sdkVersion) =>
-    VersionConstraint.parse(sdkVersion).allowsAny(_dart2);
+bool isDart2([String sdkVersion]) =>
+    VersionConstraint.parse(sdkVersion ?? 'any').allowsAny(_dart2);
```
Comment (pq) [inline lib/src/filters.dart:6]: 👍 

thanks!

**Review by pq — COMMENTED** [review]
looks awesome!

i've been sorting members to cut down on commit noise... would you consider doing that?

in IDEA:

image

File: lib/src/commands/fetch.dart (hunk 2 above)
Comment (ditman) [reply lib/src/commands/fetch.dart:142]: Yes, but it was printed 3x and it bugged me a little bit; this is what it looks like now:

```
Skipped package: marker_prof (does not use flutter)
Skipped package: handle_system (does not use flutter)
9159 packages processed
(Processed all available packages on pub.)
---
Creating: third_party/download
Extracted third_party/cache/basic_components-0.0.4+3
Installing dependencies for basic_components
```

File: lib/src/cache.dart (hunk 1 above)
Comment (ditman) [reply lib/src/cache.dart:95]: Milliseconds for this was too small, so in order for people not to pass huge numbers, the timeout is in **seconds** (it's also documented in the `--help` of the `fetch` operation)

(I could convert s -> ms for this method but... this way I report back in the error the same value the user passed in)

File: lib/src/filters.dart (hunk 3 above)
Comment (ditman) [reply lib/src/filters.dart:6]: No problem! Not sure why this is getting a null value though, maybe some bogus metadata when downloading?

**Comment by ditman** [comment]
Sorted members (of the files I touched), in vscode:

Ctrl + Shift + P (Command Palette):

Screen Shot 2019-11-05 at 2 36 52 PM

File: lib/src/commands/fetch.dart (hunk 2 above)
Comment (pq) [reply lib/src/commands/fetch.dart:142]: LGTM!

File: lib/src/cache.dart (hunk 1 above)
Comment (pq) [reply lib/src/cache.dart:95]: Ah, k.  Sounds good!

**Pull request merged**